fms/testing/fms_hf_equivalence.py

```diff
@@ -0,0 +1,314 @@
+import argparse
+import json
+import logging
+import os
+import sys
+from collections import OrderedDict
+from dataclasses import dataclass, field
+from typing import List
+
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from fms.models import get_model
+from fms.models.hf.utils import to_hf_api
+
+
+_device = "cuda"
+_decice_for_diff = "cpu"
+
+dtype_mapping = {"float16": torch.float16, "float32": torch.float32}
+
+batch_size = 8
+max_length = 2048
+
+
+def generate_and_write_tokens(
+    model, tokenizer, input_json_path, output_dir, model_impl_type
+):
+    """
+    This function loads the prompts, generate the tokens for the given prompts using the supplied model and tokenizer.
+    Finally, writes the generated tokens in a file "output_dir/model_impl_type/tokens" The output files contains serialized
+    ordered dict that contains two entries for each prompt, one for the input tokens and one for the generated tokens.
+    Paddings are removed before writing the tokens so it's safe to compare the outputs of two different runs with different batch sizes.
+
+     Parameters
+    ----------
+    model: Either of the HF or HF adapted FMS model
+    tokenizer: The tokenizer associated with the model
+    input_json_path: File path to the prompt file. The prompt file is expected to contain a set of JSON records with one attribute - "prompt".
+    output_dir: Path to the output directory
+    model_impl_type: One of the "fms" or "hf"
+    """
+    prompts = load_prompts(input_json_path)
+    logging.info(f"model = {model}")
+    logging.info(f"model config = {model.config}")
+    logging.info(f"Num prompts = {len(prompts)}")
+
+    tokenizer.pad_token = tokenizer.eos_token
+    model_pad_token_id = None
+    if hasattr(model.config, "pad_token_id") and model.config.pad_token_id != None:
+        model_pad_token_id = model.config.pad_token_id
+    else:
+        model_pad_token_id = model.config.eos_token_id
+
+    logging.debug(f"tokenizer_pad_token_id = {tokenizer.pad_token_id}")
+    logging.debug(f"model_pad_token_id = {model_pad_token_id}")
+
+    i = 0
+    global batch_size
+    data = OrderedDict()
+
+    while len(prompts) > 0:
+        if len(prompts) >= batch_size:
+            batch = prompts[:batch_size]
+            prompts = prompts[batch_size:]
+        else:
+            batch = prompts
+            prompts = []
+
+        model_inputs = tokenizer(batch, return_tensors="pt", padding=True).to(
+            device=_device
+        )
+        output = model.generate(**model_inputs, max_length=max_length)
```

Comment:
do we want to include more options for generate here that a user can pass?

Reply:
Yes, makes sense. Can do that in the next iteration in a separate PR, once we decide if this utility going to be useful.